src/ansi-c/goto-conversion/goto_clean_expr.cpp

```diff
@@ -16,9 +16,172 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <util/pointer_expr.h>
 #include <util/std_expr.h>
 #include <util/symbol.h>
+#include <util/mathematical_expr.h>
+#include <analyses/natural_loops.h>
+#include <unordered_map>
+#include <util/replace_symbol.h>
+#include <util/simplify_expr.h>
 
 #include "destructor.h"
 
+static symbol_exprt find_base_symbol(const exprt &expr)
+{
+  if(expr.id() == ID_symbol)
+  {
+    return to_symbol_expr(expr);
+  }
+  else if(expr.id() == ID_member)
+  {
+    return find_base_symbol(to_member_expr(expr).struct_op());
+  }
+  else if(expr.id() == ID_index)
+  {
+    return find_base_symbol(to_index_expr(expr).array());
+  }
+  else if(expr.id() == ID_dereference)
+  {
+    return find_base_symbol(to_dereference_expr(expr).pointer());
+  }
+  else
+  {
+    throw "unsupported expression type for finding base symbol";
+  }
+}
+
+static exprt convert_statement_expression(
+  const quantifier_exprt &qex,
+  const code_expressiont &code,
+  const irep_idt &mode,
+  goto_convertt &converter)
+{
+  goto_programt where;
+  converter.goto_convert(code, where, mode);
+  where.compute_location_numbers();
+
+  natural_loops_mutablet natural_loops(where);
+  INVARIANT(
+    natural_loops.loop_map.size() == 0,
+    "quantifier must not contain side effects");
+
+  // `last` is the instruction corresponding to the last expression in the
+  // statement expression.
+  goto_programt::const_targett last;
+  for(goto_programt::const_targett it = where.instructions.begin();
+      it != where.instructions.end();
+      ++it)
+  {
+    // `last` is an other-instruction.
+    if(it->is_other())
+    {
+      last = it;
+    }
+  }
+  auto last_expr =
+    expr_try_dynamic_cast<code_expressiont>(last->get_other())->expression();
+
+  // `pathes` contains all the `targett` we are iterating over.
+  std::vector<goto_programt::const_targett> pathes;
+  pathes.push_back(where.instructions.begin());
+  std::vector<std::pair<exprt, replace_symbolt>> path_conditions_and_value_maps;
+  path_conditions_and_value_maps.push_back(
+    std::make_pair(true_exprt(), replace_symbolt()));
```

Comment:
1. Isn't it the case that the two vectors must have a matching number of entries? I suggest you introduce a `struct` with three members and have a single vector holding elements of that struct.
2. RAII: `std::vector<goto_programt::const_targett> pathes{1, where.instructions.begin()};`